R/standardize_treatment.R

```diff
@@ -0,0 +1,445 @@
+                                                lowlim, uplim, exact_global, verbose)
+  } else {
+    constraints <- data_in$constraints
+    constraints$l[(J + 1):(J + n)] <- lowlim
+    constraints$u[(J + 1):(J + n)] <- uplim
+  }
+
+  # set optimization settings
+  if(verbose) {
+    settings <- osqp::osqpSettings(verbose = TRUE, eps_abs = 1e-6,
+                                   eps_rel = 1e-6, max_iter = 5000)
+  } else {
+    settings <- osqp::osqpSettings(verbose = FALSE, eps_abs = 1e-6,
+                                   eps_rel = 1e-6, max_iter = 5000)
+  }
+  settings <- do.call(osqp::osqpSettings,
+                      c(list(verbose = verbose,
+                             eps_rel = eps_rel,
+                             eps_abs = eps_abs),
+                        list(...)))
+
+  # solve optimization problem (possibly with uniform weights)
+  if(init_uniform) {
+    if(verbose) message("Initializing with uniform weights")
+    unifw <- get_uniform_weights_treatment(X0s, Xtaus)
+    obj <- osqp::osqp(P, q, constraints$A,
+                      constraints$l, constraints$u, pars = settings)
+    obj$WarmStart(x = unifw)
+    solution <- obj$Solve()
+  } else {
+    solution <- osqp::solve_osqp(P, q, constraints$A,
+                                 constraints$l, constraints$u,
+                                 pars = settings)
+  }
+
+  # convert weights into a matrix
+  if(verbose) message("Reordering weights...")
+  weights <- matrix(0, ncol = J, nrow = n)
+  cumsumnj <- cumsum(c(1, nj))
+  for(j in 1:J) {
+    weights[idxs[[j]], j] <- solution$x[cumsumnj[j]:(cumsumnj[j + 1] - 1)]
+  }
+
+  # compute imbalance matrix
+  imbalancetau <- as.matrix(target - t(Xtau) %*% (weights * pro_trt))
+  imbalance0 <- as.matrix(t(X0) %*% (weights * pro_trt) -
+                          t(X0) %*% (weights * pro_ctr))
+
+  # package program components if requested by user
+  if(return_program) {
+    program <- list(P = P  - lambda * I0,
+                     q = q, constraints = constraints)
+  } else {
+    program <- NULL
+  }
+
+  # return output
+  return(list(weights = weights, imbalance_0 = imbalance0,
+              imbalance_tau = imbalancetau, program = program))
+
+}
+
+
+#' Create diagonal regularization matrix
+#' @param pro_trt_split List of J vectors of treatment propensity multipliers
+#' @param pro_ctr_split List of J vectors of control propensity multipliers
+#' @param scale_sample_size Whether to scale the dispersion penalty by the sample size of each group, default T
+#' @param nj Number of units in each group
+#' @param n Total number of units
+#' @param aux_dim Dimension of auxiliary weights
+create_I0_matrix_treatment <- function(pro_trt_split, pro_ctr_split, scale_sample_size, nj, n, aux_dim) {
+
+  if(scale_sample_size) {
+    # diagonal matrix of inverse propensity scores scaled by group sample size
+    I0 <- Matrix::Diagonal(n, rep(nj ^ 2, nj) * (unlist(pro_trt_split) + unlist(pro_ctr_split)))
```

Comment:
This corresponds to the pscore scaling right?

Reply:
No, this corresponds to scaling the penalty so that greater weight is given to units from larger sites. For example, suppose that unit A is from a large site, and unit B is from a small site. If they both have the same propensity score, then they will be given equal weight if `scale_sample_size == FALSE`, and unit A will be given greater weight than unit B if `scale_sample_size == TRUE`.

Reply:
Right, but `unlist(pro_trt_split) + unlist(pro_ctr_split)` is the pscore scaling in the regularization?